### ISSUE TITLE
Comments in Config files get stripped out

### ISSUE THREAD

**Issue by duncanmcclean** [Stillat/proteus#8]
Hey,

I'm building an upgrade command for my Statamic addon, Simple Commerce to upgrade between versions. In that process, it uses Proteus to update the structure of the user's config file.

I've noticed that if I look at the config file after Proteus has made its changes, the big text code comments get stripped out. It looks like normal `// code comments` do not get stripped out, they remain in the file. I'm guessing this is probably an issue from somewhere down the stack?

Here's an example of the code comments I'm talking about:

```php
/*
    |--------------------------------------------------------------------------
    | Sites
    |--------------------------------------------------------------------------
    |
    | For each of your Statamic sites, you can setup a new store which allows you
    | to use different currencies, tax rates and shipping methods.
    |
    */
```

Here's the [full config file](https://github.com/doublethreedigital/simple-commerce/blob/master/config/simple-commerce.php) if you're interested.

**Comment by JohnathonKoster** [comment]
Good morning!

I am having troubles replicating this; could I get some a few extra details?

1. The version of `nikic/php-parser` within the `composer.lock` file and PHP version
2. Which operations you are using to update the structure of the config file
3. The "before" config file
4. The expected "after" config file

The first and second one are the most important, as some actions can work together in strange and unexpected ways.

Thank you!

**Comment by duncanmcclean** [comment]
Sure!

1. `v4.10.4`
2. Here's my code:

```php
ConfigWriter::write('simple-commerce.cart', [
  'driver' => \DoubleThreeDigital\SimpleCommerce\Orders\Cart\Drivers\SessionDriver::class,
  'key' => Config::get('simple-commerce.cart_key'),
]);

ConfigWriter::edit('simple-commerce')->remove('cart_key')->save();
```

3. [Expected before](https://github.com/doublethreedigital/simple-commerce/blob/master/config/simple-commerce.php)
4. [Expected after](https://github.com/doublethreedigital/simple-commerce/blob/v2.2-dev/config/simple-commerce.php)

**Comment by JohnathonKoster** [comment]
Looking at the before, and the expected after, the comments would not come through because the `cart` key is not present in the original, and it's formatting would not be analyzed.

I am working on something for you atm, and will provide documentation for your use case.

**Comment by duncanmcclean** [comment]
Ah okay, thanks for help (especially on Boxing Day)!

**Comment by JohnathonKoster** [comment]
@damcclean 

I've released update v1.0.6, which adds some new features that should help you out here. Because of the type of update you will be doing, we will "open" an editor instance, which allows us to specify more complicated replacements, as well as perform multiple actions at once without an IO penalty for each action.

The `edit` method simply returns a wrapper class that provides the interaction between the underlying ConfigUpdater, and Laravel's configuration repository implementation. The method we will make use of is the `replaceStructure` method, which accepts an existing key, a string to rename that key to, the new value, and a Laravel-style comment that will be used when regenerating the document. I've only tested this comment replacement with comments that have the same number of lines, and adjusting things "too much" may have unpredictable results.

```php
<?php

use Stillat\Proteus\Support\Facades\ConfigWriter;

// Open an editor instance for the simple-commerce configuration namespace.
$editInstance = ConfigWriter::edit('simple-commerce');

// Specify what the new Laravel-style comment block should be.
$newComment = <<<BLOCK
/*
|--------------------------------------------------------------------------
| Cart
|--------------------------------------------------------------------------
|
| Configure the Cart Driver in use on your site. It's what stores/gets the
| Cart ID from the user's browser on every request.
|
*/
BLOCK;

// This does a few things:
//    1. Renames the "cart_key" entry to "cart"
//    2. Replaces the existing value with whatever we supply as the 3rd argument
//    3. Rewrites the existing Laravel-style comment with the new one.
//    4. The last parameter (forceNewLine) ensures that there is an empty line
//       between the Laravel-style comment and the array key.
//
// This method will preserve the relative placement of the configuration key in the file.
$editInstance->replaceStructure('cart_key', 'cart', [
    'driver' => \DoubleThreeDigital\SimpleCommerce\Orders\Cart\Drivers\SessionDriver::class,
    'key' => 'new-key-value'
], $newComment, true);

// You may also preview the changes:
// $document = $editInstance->preview();

// Replace an existing node:
// $editInstance->replace('existing_key', ['new => 'value']);

// Change an existing value:
// $editInstance->change('existing_key', 'new-value');

// Merge new array values with existing values for a given key:
// $editInstance->merge('existing_key', ['new', 'values']);

// Completely remove an existing key:
// $editInstance->remove('existing_key');

// Persist to disk.
$editInstance->save();

```

Let me know how this works out, or if further reworks will be needed. Thanks!

**Comment by duncanmcclean** [comment]
That's perfect! Thanks for your help and providing a solution. I'll close the issue now that there's a solution!
